### DISCUSSION TITLE
5/16 課題2 コピーキャプチャと参照キャプチャ

### DISCUSSION THREAD

**Discussion by Fukky007** [jsk-lecture/software2#5] [Q&A]
03213003 福田光輝です。

上記の課題2について、
（画面左）コピーキャプチャだとうまく行くのに
（画面右）参照キャプチャだとbalanceが全て0になってしまいます。
これはどういった原理で発生しているのでしょうか。個人的には参照キャプチャでもうまく行きそうな気がしているのですが…。
初歩的な質問でしたら申し訳ありません、よろしくお願いします。

![kadai2_copy_sansho](https://user-images.githubusercontent.com/101580994/169694525-a749288e-570b-45c4-bda5-be1311773e24.png)

**Comment by k-okada** [comment]
おそらく，，になってしまいますが，以下のリンクで議論されている内容ではないか，と言う結論になりました．
https://stackoverflow.com/questions/40761035/are-c-lambdas-true-closures-capturing-by-reference

クロージャーの様に見えているけど，実際には`balance` の値は関数が変えるとともに消えてしまう（`balance`のために確保されたメモリが解放される)ので，参照キャプチャするとその部分のデータがなくなっており，`0` が表示されています．実際に何が表示されるかはプログラム中のメモリの状態次第なので，例えば`g++` の代わりに`clang --std=c++14` として実行すると毎回値が変わりました．

**Reply by Fukky007** [reply]
なるほど、関数の呼び出し終了でbalanceの元の値を格納していたメモリが消えてしまい、想定していた値と違うものが出てきてしまうということなのですかね。。コピーキャプチャだと保存されるので値を保持したまま計算できるということですか。
ご回答ありがとうございます！
